Function parameter types in macros

stepancheg opened the pull request:
`starlark_module!` macro function definitions can now optionally
specify function parameter type, and Starlark implementation will
take care of boring parameter conversion.

E. g. conversion of `Value` to `Vec<String>` previously required
explicit iteration over `Value` and downcasting each element to
`String`.

It is a simple type annotation now, and Starlark Rust embedding
should be much easier.

Example module definition:

```
starlark_module! { global =>
    cc_binary(name: String, srcs: Vec<String> = Vec::new()) {
        ...
    }
}
```

damienmg commented on starlark/src/stdlib/macros/mod.rs:
```diff
@@ -93,76 +143,84 @@ macro_rules! starlark_signature_extraction {
-    ($args:ident $call_stack:ident $env:ident $t:ident = $e:expr, $($rest:tt)* ) => {
-        starlark_signature_extraction!($args $call_stack $env $t = $e);
+    ($args:ident $call_stack:ident $env:ident $t:ident $(: $pt:ty)? = $e:expr, $($rest:tt)* ) => {
+        starlark_signature_extraction!($args $call_stack $env $t $(: $pt)? = $e);
```
In this file below there is a documentation on how to use those macro. Maybe update it a bit with the new usages?

stepancheg commented:
Rebased and added this example:

```
///     // Functions can optionally specify parameter types after colon.
///     // Parameter can be any type which implements `TryParamConvertFromValue`.
///     // When parameter type is not specified, it is defaulted to `Value`
///     // for regular parameters, `Vec<Value>` for `*args`
///     // and `LinkedHashMap<String, Value>` for `**kwargs`.
///     sqr(x: i64) {
///         Ok(Value::new(x * x))
///     }
```

damienmg commented:
> Rebased and added this example:
> 
> ```
> ///     // Functions can optionally specify parameter types after colon.
> ///     // Parameter can be any type which implements `TryParamConvertFromValue`.
> ///     // When parameter type is not specified, it is defaulted to `Value`
> ///     // for regular parameters, `Vec<Value>` for `*args`
> ///     // and `LinkedHashMap<String, Value>` for `**kwargs`.
> ///     sqr(x: i64) {
> ///         Ok(Value::new(x * x))
> ///     }
> ```

I don't see this part in the PR (wrong push?). Anyway, with that doc I think the PR is good to go so I let you update with the comment and submit it.

damienmg commented:
Ignore that I found it. Maybe some github or @damienmg's weirdness :p
